dafny: install z3 as resource

davidcok opened the pull request:
Dafny currently works with the older Z3 executable 4.8.4. There is an open issue to upgrade to the latest Z3, that work is not complete. Meanwhile, the brew download is broken and needs after the fact patching of the Z3 version. This PR fixes that problem.

gromgit commented:
Why is this needed?

davidcok commented:
Because the application was using the wrong version of Z3 — it is only capable of using the specific 4.8.4 version at present.

> On Dec 4, 2020, at 8:47 PM, Adrian Ho <notifications@github.com> wrote:
> 
> 
> Why is this needed?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Homebrew/homebrew-core/pull/66278#issuecomment-739103860>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABITDQHS3RP7G4PD6Z5T5GDSTGGJXANCNFSM4UOFNZRA>.
>

carlocab commented on Formula/dafny.rb:
```diff
@@ -37,10 +29,15 @@ def install
 
+    cd libexec do
+      system "curl", "-sSL", "-o", "file.zip", "https://github.com/Z3Prover/z3/releases/download/z3-4.8.4/z3-4.8.4.d6df51951f4c-x64-osx-10.14.1.zip"
+      system "unzip", "-q", "file.zip"
```
This should be done using a `resource` block, not through curl. `resource` blocks check source integrity by verifying `sha256`.

Have a look at another formula for examples. `grep  'resource do'` in the core repo would work.

carlocab replied:
Also, this looks like it downloads something that is pre-built, which Homebrew doesn't like, I think.

davidcok replied:
Fixed.

carlocab commented on Formula/dafny.rb:
```diff
@@ -3,28 +3,25 @@ class Dafny < Formula
-    sha256 "4b0bb8f5e2385b99318cc85ff38496e87814d5658f5dd4054fdc7d2a0a8ebc07" => :mojave
-    sha256 "4b64f7c46ab2fdfb997ca918c81e0d423609f565bdd405eaa9a2d7e848295ab7" => :high_sierra
-  end
-
```
You're not supposed to touch the `bottle` block.

davidcok replied:
Put it back

carlocab commented on Formula/dafny.rb:
```diff
@@ -3,28 +3,25 @@ class Dafny < Formula
   resource "boogie" do
     url "https://github.com/boogie-org/boogie.git",
-      revision: "9e74c3271f430adb958908400c6f6fce5b59000a"
+    revision: "9e74c3271f430adb958908400c6f6fce5b59000a"
```
Please don't remove the indentation. Indenting makes it clear that the `revision:` line is a continuation of the previous line.

davidcok replied:
OK

SMillerDev commented:
Is it actually broken with this version? Or is it just unsupported?

carlocab commented:
It hasn't been bottled for Big Sur, so I'm going to guess that it's actually broken.

carlocab commented on Formula/dafny.rb:
```diff
@@ -51,9 +52,16 @@ def install
       }
     EOS
-    system "#{bin}/dafny", testpath/"test.dfy"
+    assert_equal "Dafny 2.3.0.10506\n\nDafny program verifier finished with 1 verified, 0 errors\n",
```
Is there any way you could write this test without hard-coding version information? This makes the formula difficult to maintain.

davidcok replied:
Removed the version from the output using /nologo

davidcok commented:
It does not work with newer versions — we know it needs fixing, but we need this change to the formula to have the download work out of the box.

> On Dec 5, 2020, at 7:23 AM, Sean Molenaar <notifications@github.com> wrote:
> 
> 
> Is it actually broken with this version? Or is it just unsupported?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Homebrew/homebrew-core/pull/66278#issuecomment-739244136>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABITDQEEOMS4FOSTY6ORQIDSTIQ4NANCNFSM4UOFNZRA>.
>

carlocab commented:
Will dafny support the latest z3 in the next version?

davidcok commented:
Probably not We hope to have a new 3.0 version out in January (and a new pre-release momentarily), but upgraded Z3 support is not in the plan for that milestone.

This fix is required to support current issues reported by users.

> On Dec 5, 2020, at 10:49 AM, Carlo Cabrera <notifications@github.com> wrote:
> 
> 
> Will dafny support the latest z3 in the next version?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Homebrew/homebrew-core/pull/66278#issuecomment-739312947>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABITDQGR74JJ5QYFDIKKMZLSTJJAHANCNFSM4UOFNZRA>.
>

davidcok commented:
I have addressed all the PR comments on PR66278.
However, as pointed out the CI run fails for Catalina. But this appears to be in the setup of the homebrew environment, and not in anything related to the Dafny brew formula.
Some advice on this point would be much appreciated.

Thanks for any help,
- David


> On Dec 5, 2020, at 10:49 AM, Carlo Cabrera <notifications@github.com> wrote:
>

carlocab commented:
The CI nodes are acting up at the moment, so we'll have to wait for them to be fixed.

In the meantime, it would be useful if you could leave a comment before the `z3` `resource` block explaining:
1. how and when to update it;
2. when it can be replaced with a dependency on the `z3` formula (a link to a tracking issue would be enough)

This will be useful for anyone else who tries to update the `dafny` formula.

Apologies for making comments on a piecemeal basis -- I know it'd have been a lot more convenient had I put all my comments together. I promise that that's the last comment you'll get out of me.

davidcok commented:
Done. And no problem. I have long used brew on my (various) Macs, but this is the first time managing a brew formula — so I appreciate the careful reviews.

- David

> On Dec 7, 2020, at 4:36 PM, Carlo Cabrera <notifications@github.com> wrote:
> 
> 
> The CI nodes are acting up at the moment, so we'll have to wait for them to be fixed.
> 
> In the meantime, it would be useful if you could leave a comment before the z3 resource block explaining:
> 
> how and when to update it;
> when it can be replaced with a dependency on the z3 formula (a link to a tracking issue would be enough)
> This will be useful for anyone else who tries to update the dafny formula.
> 
> Apologies for making comments on a piecemeal basis -- I know it'd have been a lot more convenient had I put all my comments together. I promise that that's the last comment you'll get out of me.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Homebrew/homebrew-core/pull/66278#issuecomment-740194341>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABITDQCI7UVHOKZB4JSM4K3STVDFBANCNFSM4UOFNZRA>.
>

carlocab commented on Formula/dafny.rb:
```diff
@@ -20,13 +20,22 @@ class Dafny < Formula
   depends_on "nuget" => :build
   depends_on "mono"
-  depends_on "z3"
+  # Use the following when dafny uses the latest z3 (and omit the 
```
```suggestion
  # Use the following when dafny uses the latest z3 (and omit the
```
CI is failing because of the trailing whitespace here.

davidcok replied:
Thanks. Fixed.

davidcok commented:
Thanks, Fixed. And tests passed. I hope you can approve/merge this.

- David


> On Dec 10, 2020, at 3:06 PM, Carlo Cabrera <notifications@github.com> wrote:
> 
> 
> @carlocab commented on this pull request.
> 
> In Formula/dafny.rb <https://github.com/Homebrew/homebrew-core/pull/66278#discussion_r540461547>:
> 
> > @@ -20,13 +20,22 @@ class Dafny < Formula
>    depends_on "mono-libgdiplus" => :build
>    depends_on "nuget" => :build
>    depends_on "mono"
> -  depends_on "z3"
> +  # Use the following when dafny uses the latest z3 (and omit the 
> ⬇️ Suggested change
> -  # Use the following when dafny uses the latest z3 (and omit the 
> +  # Use the following when dafny uses the latest z3 (and omit the
> CI is failing because of the trailing whitespace here.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Homebrew/homebrew-core/pull/66278#pullrequestreview-549559784>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABITDQADRXRCHOPJT6M3DITSUES4HANCNFSM4UOFNZRA>.
>

carlocab commented:
We'll need @SMillerDev for that. Thanks for the contribution.

SMillerDev commented:
Could you rebase this pull request @davidcok ?

davidcok commented:
Done - and checks passed.

fxcoudert commented:
@davidcok  We're going to need this to be rebased as a single commit on top of master — and no merge commits. That's how our build-bot can handle and merge automatically, including the newly built binaries.

davidcok commented:
Rebased and checks passed.
